compiler/hash-pipeline/src/args.rs

```diff
@@ -0,0 +1,252 @@
+    let mut settings = CompilerSettings::default();
+    let mut args = std::env::args().skip(1);
+
+    while let Some(arg) = args.next() {
+        // This is a configuration key that specifies the "key" and then
+        // the value in the form of `-C<key>=<value>`
+        if arg.starts_with("-C") || arg.starts_with("--") {
+            parse_option(&mut settings, &mut args, arg.as_str())?;
+        } else {
+            // This is specifying what kind of a stage the compiler should run
+            // the job on whether it is `build`, `check`,
+            // `ast-gen`...
+            match arg.as_str() {
+                "build" => {
+                    settings.stage = CompilerStageKind::Full;
+                }
+                "check" => {
+                    settings.stage = CompilerStageKind::Typecheck;
+                }
+                "ast-gen" => {
+                    settings.stage = CompilerStageKind::Parse;
+                }
+                "ir-gen" => {
+                    settings.stage = CompilerStageKind::IrGen;
+                }
+                _ => {
+                    return Err(ArgumentError::UnknownStage(arg));
+                }
+            };
+
+            // The next argument after this is the input file.
+            if let Some(filename) = args.next() {
+                settings.entry_point = Some(filename);
+            } else {
+                return Err(ArgumentError::MissingEntryPoint);
+            }
+        }
+    }
+
+    Ok(settings)
+}
+
+/// This function is used to parse a single option that is passed to the
+/// compiler. The option can either be in the form of a configuration key
+/// `-C<key>=<value>` or a flag `--flag`.
+///
+/// N.B. This function does not deal with the entry point of the compiler.
+pub fn parse_option(
+    settings: &mut CompilerSettings,
+    args: &mut impl Iterator<Item = String>,
+    arg: &str,
+) -> Result<(), ArgumentError> {
+    // This is a configuration key that specifies the "key" and then
+    // the value in the form of `-C<key>=<value>`
+    if arg.starts_with("-C") {
+        // Check if the key is part of this argument or if there is a
+        // separate argument for the value, which we assume to be the
+        // next argument.
+        if arg.len() == 2 {
+            if let Some(arg) = args.next() {
+                parse_arg_configuration(settings, arg)?;
+            } else {
+                return Err(ArgumentError::UnknownKey(arg.to_string()));
+            }
+        } else {
+            parse_arg_configuration(settings, arg.trim_start_matches("-C").to_string())?;
+        }
+    } else if arg.starts_with("--") {
+        let key = arg.trim_start_matches("--").to_string();
+
+        match key.as_str() {
+            "debug" => {
+                settings.debug = true;
+            }
+            "output-metrics" => {
+                settings.output_metrics = true;
+            }
```

Comment:
Would be nice to have these in some macro form eventually..

Reply:
Yes, it would be nice, and it could capture the type of argument, some kind of description, and potentially some parsing parameters. Ill add a `@@Todo`...